[OCPCLOUD-1248] Add initial iteration of ControlPlaneMachineSet operator enhancement (#1008)
openshift/enhancements: enhancements/machine-api/control-plane-machine-set.md

    // provider spec.
    UpdatedReplicas int32 `json:"updatedReplicas,omitempty"`

    // UnavailableReplicas is the number of Control Plane Machines that are
    // still required before the ControlPlaneMachineSet reaches the desired
    // available capacity. When this value is non-zero, the number of
    // ReadyReplicas is less than the desired Replicas.
    UnavailableReplicas int32 `json:"unavailableReplicas,omitempty"`

    // Conditions represents the observations of the ControlPlaneMachineSet's current state.
    // Known .status.conditions.type are: (TODO)
    // TODO: Identify different condition types/reasons that will be needed.
    // +patchMergeKey=type
    // +patchStrategy=merge
    // +listType=map
    // +listMapKey=type
    // +optional
    Conditions []metav1.Condition `json:"conditions,omitempty"`
}

// FailureDomain represents the different configurations required to spread Machines
// across failure domains on different platforms.
// +union
type FailureDomains struct {
  // Platform identifies the platform for which the FailureDomain represents
  // +kubebuilder:validation:Enum:="aws";"azure";"gcp";"openstack"
  // +unionDiscriminator
  // +optional

>   Platform configv1.PlatformType `json:platform,omitempty`

Author: API reviewer question: Because I've made this `configv1.PlatformType`, it should be one of the platform type consts, the issue here is that this list is larger than we support right now (eg include Nutanix and other platforms that won't initially be support), the other issue is that the constants of this type are camel case, so for example `AWS` instead of `aws`, `OpenStack` not `openstack`.

IIUC, the discriminator needs to match the json value of the fields in the union, so we need the lower case versions to match the json tags? Is that correct?

If it's correct, I guess I switch this back to a string and keep the kubebuilder enum validation

  unsupported by OpenShift)
* Management of the etcd cluster state (the etcd operator will handle this separately)
* Automatic adoption of existing clusters (we will provide instructions to allow users to opt-in for existing cluster)
* Management of anything that falls out of the scope of Machine API

## Proposal

A new CRD `ControlPlaneMachineSet` will be introduced into OpenShift and a respective `control-plane-machine-set-
operator` will be introduced as a new second level operator within OpenShift to perform the operations described in
this proposal.

The new CRD will define the specification for managing (creating, adopting, updating) Machines for the Control Plane.
The operator will be responsible for ensuring the desired number of Control Plane Machines are present within the
cluster as well as providing update mechanisms akin to those seen in a StatefulSets and Deployments to allow rollouts
of updated configuration to the Control Plane Machines.

### User Stories

- As a cluster administrator of OpenShift, I would like to be able to safely and automatically vertically resize my
  control plane as and when the demand on the control plane changes

> - As a cluster administrator of OpenShift, I would like to be able to automatically recover failed Control Plane

Author: Ahh right, this is in terms of a `Failed` phase machine, so normally this would be that the instance has been removed from the cloud provider for some reason, or possibly a MachineHealthCheck has determined that it's unhealthy and should be replaced, I will clarify

}

// FailureDomain represents the different configurations required to spread Machines
// across failure domains on different platforms.
// +union
type FailureDomain struct {
  // Platform identifies the platform for which the FailureDomain represents
  // +kubebuilder:validation:Enum:="aws";"azure";"gcp";"openstack"
  // +unionDiscriminator
  // +optional
  Platform configv1.PlatformType `json:platform,omitempty`

  // AWS configures failure domain information for the AWS platform
  // +optional
  AWS *AWSFailureDomain `json:aws,omitempty`

  // Azure configures failure domain information for the Azure platform
  // +optional
  Azure *AzureFailureDomain `json:azure,omitempty`

  // GCP configures failure domain information for the GCP platform
  // +optional
  GCP *GCPFailureDomain `json:gcp,omitempty`

  // OpenStack configures failure domain information for the OpenStack platform
  // +optional
  OpenStack *OpenStackFailureDomain `json:openstack,omitempty`
}

// AWSFailureDomain configures failure domain information for the AWS platform
type AWSFailureDomain struct {
  // Subnet is a reference to the subnet to use for this instance
  // +optional

> 	Subnet AWSResourceReference `json:"subnet,omitempty"`

Reviewer: I think there might be mixed indentation here. And in a couple of places below too.

Author: Good catch

  // Placement configures the placement information for this instance

>   // +optiona;

Reviewer: ```suggestion
  // +optional
```

cluster) and as such, we see that HA clusters will continue to be used and we must solve this problem to allow the
continued adoption of HA clusters.

### Goals

* Provide a "best practices" approach to declarative management of Control Plane Machines
* Allow users to "1-click" scale their Control Plane to large or smaller instances
* Allow the adoption of existing Control Planes into this new mechanism
* Provide a safe mechanism to perform the sensitive operation of scaling the Control Plane and provide adequate
  feedback to end users about progress of the operation
* Allow users to opt-out of Control Plane management should our mechanism not fit their needs
* Allow users to customise rolling update strategies based on the needs of their environments

### Non-Goals

* Allow horizontal scaling of the Control Plane (this may be required in the future, but today is considered
  unsupported by OpenShift)
* Management of the etcd cluster state (the etcd operator will handle this separately)
* Automatic adoption of existing clusters (we will provide instructions to allow users to opt-in for existing cluster)

> * Management of anything that falls out of the scope of Machine API

Author: I think one thing that was mentioned before was the load balancers for the control plane instance, MAPI manages the membership but it doesn't manage the load balancers themselves, I'll add this as an additional note

### User Stories

- As a cluster administrator of OpenShift, I would like to be able to safely and automatically vertically resize my
  control plane as and when the demand on the control plane changes
- As a cluster administrator of OpenShift, I would like to be able to automatically recover failed Control Plane
  Machines (eg those that have been removed by the cloud provider, or those failing MachineHealthChecks)
- As a cluster administrator of OpenShift, I would like to be able to make changes to the configuration of the
  underlying hardware of my control plane and have these changes safely applied using immutable hardware concepts
- As a cluster administrator of OpenShift, I would like to be able to control rollout of changes to my Control Plane
  Machines such that I can test changes with a single replica before applying the change to all replicas
- (Future work) As a cluster administrator of OpenShift with restricted hardware capacity, I would like to be able to   
  scale down my control plane before adding new Control Plane Machines with newer configuration, notably, my
  environment does not have capacity to add additional Machines during updates

### API Extensions

> We will introduce a new `ControlPlaneMachineSet` CRD to the `machine.openshift.io/v1b` API group. It will be based

Author: `machine.openshift.io/v1`, the `b` is a typo

on the spec and status structures defined below.

```go
// ControlPlaneMachineSet represents the configuration of the ControlPlaneMachineSet.
type ControlPlaneMachineSetSpec struct {
	// Replicas defines how many Control Plane Machines should be
	// created by this ControlPlaneMachineSet.
	// This field is immutable and cannot be changed after cluster
	// installation.
	// The ControlPlaneMachineSet only operates with 3 or 5 node control planes,
	// 3 and 5 are the only valid values for this field.
	// +kubebuilder:validation:Enum:=3;5
	// +kubebuilder:default:=3
	// +kubebuilder:validation:Required

> 	Replicas *int32 `json:"replicas"`

Reviewer 2: Why is this a pointer if it is a required field?

Author: I believe there were some comments on this before, IIRC this is because we are using a scale subresource and with other resources that use the scale subresoruce, the spec replicas are expected to be a pointer. To ensure consistency between this and other scalable resources, we kept this one as a pointer

Author: Yeah this is 100% a future proofing action. It's likely we will want to support scale in the future from conversations I have had, but for now we aren't going to. The ill-effect of having this as a pointer is minimal given the issues we would face if we wanted to change it from a reference to a pointer later IMO, so better to keep as a pointer for now to be consistent and future proof ourselves

Author: The MachineSpec is within the OpenShiftMachineV1Beta1Template in this API, we include a `machinev1beta1.MachineSpec` as `spec` which is the template of the Machine we use to stamp out new Machines. 

If the user changes the providerSpec, eg the vmSize, we will perform a roll out for them

)

// OpenShiftMachineV1Beta1MachineTemplate is a template for the ControlPlaneMachineSet to create
// Machines from the v1beta1.machine.openshift.io API group.
type OpenShiftMachineV1Beta1MachineTemplate struct {
	// FailureDomains is the list of failure domains (sometimes called
	// availability zones) in which the ControlPlaneMachineSet should balance
	// the Control Plane Machines.
	// This will be merged into the ProviderSpec given in the template.
	// This field is optional on platforms that do not require placement
	// information, eg OpenStack.
	// +optional
	FailureDomains FailureDomains `json:"failureDomains,omitempty"`

	// ObjectMeta is the standard object metadata
	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
	// Labels are required to match the ControlPlaneMachineSet selector.
	// +kubebuilder:validation:Required
	ObjectMeta ControlPlaneMachineSetTemplateObjectMeta `json:"metadata"`

	// Spec contains the desired configuration of the Control Plane Machines.
	// The ProviderSpec within contains platform specific details
	// for creating the Control Plane Machines.

> 	// The ProviderSe should be complete apart from the platform specific

Reviewer 3: typo?
```suggestion
	// The ProviderSpec should be complete apart from the platform specific
```

Author: That's been resolved on the merged API, need to update this before me merge this enhancement

// across failure domains on different platforms.
// +union
type FailureDomains struct {
	// Platform identifies the platform for which the FailureDomain represents
	// +unionDiscriminator
	// +optional
	Platform configv1.PlatformType `json:"platform,omitempty"`

	// AWS configures failure domain information for the AWS platform
	// +optional
	AWS *[]AWSFailureDomain `json:"aws,omitempty"`

	// Azure configures failure domain information for the Azure platform
	// +optional
	Azure *[]AzureFailureDomain `json:"azure,omitempty"`

	// GCP configures failure domain information for the GCP platform
	// +optional
	GCP *[]GCPFailureDomain `json:"gcp,omitempty"`

	// OpenStack configures failure domain information for the OpenStack platform
	// +optional
	OpenStack *[]OpenStackFailureDomain `json:"openstack,omitempty"`
}

// AWSFailureDomain configures failure domain information for the AWS platform
type AWSFailureDomain struct {
	// Subnet is a reference to the subnet to use for this instance
	// +optional
	Subnet AWSResourceReference `json:"subnet,omitempty"`

	// Placement configures the placement information for this instance
	// +optional

> 	Placement AWSFailureDomainPlacement `json:"placement,omitempty"`

Reviewer 2: Do we need the placement, as it is redundant with the subnet? The subnet is already associated with a specific availability zone.

Author: The subnet is optional in Machine API, it has the ability to lookup a default subnet based on the availability zone provided if the subnet is not provided. As such, we need to maintain this behaviour between this API and the main AWS providerSpec API

Reviewer 2: OK. It was not obvious to me from the comments that the subnet is optional when specifying placements for AWS. With the union, it is difficult to ascertain which fields are actually optional for a given platform. I recommend that we add some more text to the description of `subnet` indicating that a user can be omit it to have the operator select the appropriate subnet for the user (and perhaps the limitations of that).

Author: The kubeubilder tags for optional and required should be accurate, for example there's a tag on this struct that indicates you must set at least one of the values within it, but both are optional, which makes the openapi error at you if you provide an empty spec, but if either field is set, it will accept it.

I will expand the subnet comment though as you're right, it's a little light in detail at the moment

}

// AWSFailureDomainPlacement configures the placement information for the AWSFailureDomain
type AWSFailureDomainPlacement struct {
	// AvailabilityZone is the availability zone of the instance
	// +optional

> 	AvailabilityZone string `json:"availabilityZone,omitempty"`

Reviewer 2: What is the purpose of putting the `availabilityZone` field one level deeper than the similar `zone` fields for other platforms?

Author: This is so that it matches the design of the `AWSMachineProviderConfig` from a user perspective, we thought it would be more obvious what the effect of the failure domain is, if it were to look like a subset of the providerSpec

Reviewer 2: Makes sense.

`ControlPlaneMachineSet` to allow the controller to ensure a safe removal of the ControlPlaneMachineSet, while leaving
the Control Plane Machines in place.

Notably it will need to ensure that there are no owner references on the Machines pointing to the ControlPlaneMachineSet
instances. This will prevent the garbage collector from removing the Control Plane Machines when the
`ControlPlaneMachineSet` is deleted.

If users later wish to reinstall the `ControlPlaneMachineSet`, they are free to do so.

Note: Owner references will be added to the Control Plane Machines to identify to other components that a controller is
managing the state of these Machines. This allows other systems such as the MachineHealthCheck to identify that if they
were to make an action on the Machine, the Control Plane Machine Set Operator will react to that action.

#### Installing a ControlPlaneMachineSet within an existing cluster

When adding a ControlPlaneMachineSet to a existing cluster, the end user will need to define the
`ControlPlaneMachineSet` resource by copying the existing Control Plane Machine ProviderSpecs.

> Once this is copied, they should remove the failure domain and add the desired failure domains to the `FailureDomains`

Reviewer 3: > remove the failure domain and add the desired failure domains to the `FailureDomains`

Just a bit confused by this - where are they removing the failure domain?

Author: Yeah that's it, to give an explicit example, and I'm not sure how to word this well but:
- User copies providerSpec from an existing ControlPlaneMachine
- They paste that onto the new `ControlPlaneMachineSet.Spec.Template.machines_v1beta1_machine_openshift_io.Spec.ProviderSpec.Value`
- In the case of AWS, they now delete `Placement.AvailabilityZone` and `Subnet` from the `ProviderSpec.Value`
- They now add their failure domains to `ControlPlaneMachineSet.Spec.Template.machines_v1beta1_machine_opneshift_io.FailureDomains.AWS`
- The failure domains will look like:
```
failureDomains:
  platform: aws
  aws:
  - placement:
      availabilityZone: us-east-1a
    subnet:
      type: filters
      filters:
        name: "tag:Name"
        values:
        - subnet-us-east-1a
  - placement:
      availabilityZone: us-east-1b
    subnet:
      type: filters
      filters:
        name: "tag:Name"
        values:
        - subnet-us-east-1b
  - placement:
      availabilityZone: us-east-1c
    subnet:
      type: filters
      filters:
        name: "tag:Name"
        values:
        - subnet-us-east-1c
```

Reviewer 3: Perfect, that's very helpful!

There are however a few concerns about this idea which means we are planning to not implement this (at least for the
first iteration of the project):
- It may encourage users to attempt to use `ControlPlaneMachineSets` with UPI clusters which are incorrectly configured
  - Some users have Machines in their cluster, that aren't actually configured correctly, therefore we would be creating
    an incorrectly configured `ControlPlaneMachineSet` for them, it is then unclear who is at fault for this and we
    could end up with an increased number of support tickets due to these misconfigurations
  - As an example, this is particularly prevalent in UPI clusters where users can forget to remove the Machine manifests
    from the install manifest directory, this has been the source of many bugs where users have had Machines stuck in
    Provisioning for an extended period. Inferring the specs from these Machines would certainly make the
    `ControlPlaneMachineSet` invalid.
- Some users may have made out of band changes to their Control Plane Machines which are not reflected within the
  Machine specs

>   - We document the procedure within product docs of replacing a Control Plane Machine, though it is a manual process

Reviewer 3: I am guessing that at least for now, we will want to keep the manual steps too, and perhaps just highlight that this feature will make the process much easier for supported configurations?

Author: I think we will want to have a manual process and a, with a ControlPlaneMachineSet. And if we can cross reference so that users know about the new type in the manual process, then great, that makes things a lot easier :)

	OnDelete ControlPlaneMachineSetStrategyType = "OnDelete"
)

// FailureDomain represents the different configurations required to spread Machines
// across failure domains on different platforms.
// +union
type FailureDomains struct {
	// Platform identifies the platform for which the FailureDomain represents
	// +unionDiscriminator
	// +optional
	Platform configv1.PlatformType `json:"platform,omitempty"`

	// AWS configures failure domain information for the AWS platform
	// +optional
	AWS *[]AWSFailureDomain `json:"aws,omitempty"`

	// Azure configures failure domain information for the Azure platform
	// +optional
	Azure *[]AzureFailureDomain `json:"azure,omitempty"`

	// GCP configures failure domain information for the GCP platform
	// +optional
	GCP *[]GCPFailureDomain `json:"gcp,omitempty"`

	// OpenStack configures failure domain information for the OpenStack platform
	// +optional
	OpenStack *[]OpenStackFailureDomain `json:"openstack,omitempty"`
}

// AWSFailureDomain configures failure domain information for the AWS platform
// +kubebuilder:validation:MinProperties:=1
type AWSFailureDomain struct {
	// Subnet is a reference to the subnet to use for this instance.

>   // If no subnet reference is provided, the Machine will be created in the first
>   // subnet returned by AWS when listing subnets within the provided availability zone.

Reviewer 2: nit on spacing
```suggestion
	// If no subnet reference is provided, the Machine will be created in the first
	// subnet returned by AWS when listing subnets within the provided availability zone.
```

>   // subnet returned by AWS when listing subnets within the provided availability zone.

Reviewer 2: Is this the current behavior in the Machine API? It seems unlikely (1) that an account will only have a single subnet in a region and (2) that in the case of multiple subnets that the first one selected will be the one that the user wants to use.

Author: Yep, this is behaviour that was carried over from Cluster API when we forked from that initially, not sure this behaviour is great, but we can't change it at this point

Reviewer 2: Are we stuck on perpetuating that to the ControlPlaneMachineSet? Do we not have the choice to make this subnets field mandatory? Or does that go back to diverging too much from Cluster API, with respect to the possibility of merging back with them at some point in the future? Could we start as required and make the field optional in the future?

Author: > Are we stuck on perpetuating that to the ControlPlaneMachineSet?

I had intended to keep the behaviour the same so that we don't have to have caveats in our documentation such as "this is required for Control Plane Machines but optional for Worker Machines"

> Could we start as required and make the field optional in the future?

I've asked API folks about this in the past, changing required to optional can cause issues for API consumers. If we define an API field as required, then the consumer doesn't need to handle when it's not set, changing it to optional later would then break their logic, so this in general is not allowed

> Or does that go back to diverging too much from Cluster API, with respect to the possibility of merging back with them at some point in the future?

Cluster API have optional subnets on their Machines so ideally we stick to that too. In their case however, there are a list of subnets on the `AWSCluster` resource which they fall back to when it's not provided, so it's a little better in terms of the behaviour.

IMO at the moment at least, we should leave this optional and retain the same behaviour between control plane and worker machines. All of our docs examples include the subnets already and I would expect users to start their CPMS config based on their existing control plane machines, which will have the subnets defined. 
If they don't define the subnets, then the Machines may end up in the correct subnet, or they won't, and if they don't, they will likely never become running because the traffic between the new node and the rest of the cluster won't be allowed. In this case, our new CPMS operator will go degraded and highlight to them that there is an issue

Reviewer 2: Is the subnet set in the Machine somewhere? Are we covered in a scenario where there are subnets added to the account *after* the ControlPlaneMachineSet is created with omitted subnets? For example, if a machine needs to be re-created, can we guarantee that the new machine is added to the same subnet as the existing machines?

Reviewer 2: >  > Is the subnet set in the Machine somewhere?

> Yes

Sorry, I was not clear enough here. What I meant to ask was whether the subnet chosen by the machine-api when the user omits the subnet would be stored in the Machine resource.

> I would expect when we integrate this with the installer, it will set the subnets by filter anyway, so this is only in theory an issue for users who install this as a day 2

Right. I don't expect this to be an issue for the vast, vast majority of users. However, that again begs the question of why we would want to even offer the option to configure it in a way that has the high likelihood of failing. I get that we are trying to be consistent between APIs. There is value in that. But we should also consider whether we want to perpetuate deficiencies in an API solely for the sake of consistency. If that consideration has been made--which it seems to have been--and the conclusion is that we do prefer consistency in this case, then I can live with that. I do not share the same opinion, but that is fine.

Author: > Sorry, I was not clear enough here. What I meant to ask was whether the subnet chosen by the machine-api when the user omits the subnet would be stored in the Machine resource.

Oh got it, no we don't store it anywhere

> If that consideration has been made--which it seems to have been--and the conclusion is that we do prefer consistency in this case, then I can live with that. I do not share the same opinion, but that is fine.

It has previously been discussed within the team, but I can bring it up again in case opinions have changed. In the prior conversation I think the big sticking point was that we don't want to block a user from doing something on the control plane that they in theory can do today with an unmanaged control plane machine. If we were to do that, then there's a probably very small chance that someone might have to circumvent the management we are putting in, at which point this proposal adds negative value for them.

I think the previous conversation was around options and fields in general (and trying to create a blanket rule) rather than specifically the subnets, so I'll bring this up with the team and see if anyone has a changed opinion if we look directly at the subnet case

Author: I discussed this with the team and for now, we prefer to keep the consistency with the worker and control plane machines even in the case of the subnets, going to resolve this

```yaml
failureDomains:
  aws:
  - placement:
      availabilityZone: us-east-1a
    subnet:
      filters:
      - name: "tag:Name"
        values:
        - "my-cluster-subnet-1a"
  - placement:
      availabilityZone: us-east-1b
      subnet:
        filters:
        - name: "tag:Name"
          values:
          - "my-cluster-subnet-1b"
```

A user on Azure may set their `FailureDomains` as:

```yaml
failureDomains:
  azure:
  - zone: us-central-1
  - zone: us-central-2
```

###### Failure Domains on vSphere

As of OpenShift 4.10, there is no concept of a zone within vSphere and as such we would expect that the `FailureDomains`
would be omitted for vSphere.

However, [there is future work](https://github.com/openshift/enhancements/pull/918) to include zone support for vSphere
within 4.11. Once this is implemented, the Machine API provider for vSphere will understand the concept of zones,
defined within the `Infrastructure` resource.
The `ControlPlaneMachineSet` will then be able to balance Machines across the different zones by setting the new zone
information within the provider specs as it does on other platforms.

> #### Ensuring Machines match the desired state

Reviewer 4: @JoelSpeed Is there any reason why a control plane should not perform veritical scaling while the control plane is upgrading? My thoughts are this should be ok, just wanted to raise this question. Apologies if its been raised elsewhere.

cc @jewzaam

Author: I think this was previously discussed and we concluded that because the upgrade procedure itself is disruptive to the control plane, mixing vertical scaling into this could cause significant extra disruption and elongate both the process of upgrade and the rollout.

As these are both high risk scenarios, I think we concluded that it would be better to have them be mutually exclusive, though I can't find the previous conversation now

Reviewer 4: Thanks @JoelSpeed I think i remember discussing elsewhere as well :man_shrugging:  - is there a webhook or another implimentation to prevent triggering veritical scaling during upgrades?

Author: This will be prevented within the controller itself, ie, users will be able to make the change, but the operator will mark itself as unable to rollout due to the pending update

The operator will be responsible for ensuring the desired number of Control Plane Machines are present within the
cluster as well as providing update mechanisms akin to those seen in a StatefulSets and Deployments to allow rollouts
of updated configuration to the Control Plane Machines.

### User Stories

- As a cluster administrator of OpenShift, I would like to be able to safely and automatically vertically resize my
  control plane as and when the demand on the control plane changes
- As a cluster administrator of OpenShift, I would like to be able to automatically recover failed Control Plane
  Machines (eg those that have been removed by the cloud provider, or those failing MachineHealthChecks)
- As a cluster administrator of OpenShift, I would like to be able to make changes to the configuration of the
  underlying hardware of my control plane and have these changes safely applied using immutable hardware concepts
- As a cluster administrator of OpenShift, I would like to be able to control rollout of changes to my Control Plane
  Machines such that I can test changes with a single replica before applying the change to all replicas

> - (Future work) As a cluster administrator of OpenShift with restricted hardware capacity, I would like to be able to   

Author: The current approach is handling replacement w/scale up and down, the point of this line in particular is that currently we require surging the capacity as a part of the replacement. This story in particular is referencing a recreate style replacement rather than a rolling update style replacement

> - (Future work) As a cluster administrator of OpenShift with restricted hardware capacity, I would like to be able to   

Reviewer 5: That sounds like a limitation that should be explicitly expressed.

> - (Future work) As a cluster administrator of OpenShift with restricted hardware capacity, I would like to be able to   

Author: This is already expanded upon in the `The Recreate update strategy (FUTURE WORK)` section, but I can reword this here if you think there's a way I can make this clearer in the goals

> - (Future work) As a cluster administrator of OpenShift with restricted hardware capacity, I would like to be able to   

Author: Another note on this, this project has been designed with both cloud and local datacenter solutions in mind, though, the configuration changes we are targeting, eg resizing, don't make sense in baremetal environments. So we aren't sure whether we actually want to support this Recreate strategy ever, but we left it in for prosperity to facilitate a future conversation about this

type ControlPlaneMachineSetMachineType string

const (
	// OpenShiftMachineV1Beta1MachineType is the OpenShift Machine API v1beta1 Machine type.
	OpenShiftMachineV1Beta1MachineType ControlPlaneMachineSetMachineType = "machines_v1beta1_machine_openshift_io"
)

// OpenShiftMachineV1Beta1MachineTemplate is a template for the ControlPlaneMachineSet to create
// Machines from the v1beta1.machine.openshift.io API group.
type OpenShiftMachineV1Beta1MachineTemplate struct {
	// FailureDomains is the list of failure domains (sometimes called
	// availability zones) in which the ControlPlaneMachineSet should balance
	// the Control Plane Machines.
	// This will be merged into the ProviderSpec given in the template.
	// This field is optional on platforms that do not require placement
	// information, eg OpenStack.
	// +optional
	FailureDomains FailureDomains `json:"failureDomains,omitempty"`

	// ObjectMeta is the standard object metadata
	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
	// Labels are required to match the ControlPlaneMachineSet selector.
	// +kubebuilder:validation:Required

> 	ObjectMeta ControlPlaneMachineSetTemplateObjectMeta `json:"metadata"`

Author: As a rule in OpenShift APIs we avoid reusing upstream objects in our specs as they add fields to our APIs that we don't care about. For example, having the whole `metav1.ObjectMeta` here would mean fields like the name or creation timestamp are now a part of this API, yet they have no meaning to the controller and if a user sets them, are ignored.

It's better to have a struct with only those fields we actually care about so that the generated schema can give better feedback to the user about what is and isn't supported.

The failure domain (sometimes called availability zone) is a part of this provider spec and as such, defines that all
Machines within the Machineset share a failure domain.

This is undesirable for Control Plane Machines as we wish to have them spread across multiple availability zones to
reduce the likelihood of datacenter level faults degrading the control plane.
To this end, the failure domains for the control plane will be set on the `ControlPlaneMachineSet` spec directly.

When creating Machines, the `ControlPlaneMachineSet` controller will balance the Machines across these failure domains
by injecting the desired failure domain for the new Machine into the provider spec based on the platform specific
failure domain field.

##### Failures domains

The `FailureDomains` field is expected to be populated by the user/installer based on the topology of the Control Plane
Machines. For example, we expect on AWS that this will contain a list of availability zones within a single region.
Note, we are explicitly not expecting users to add different regions to the the `FailureDomains` as we do not support

> running OpenShift across multiple regions.

Author: If the user doesn't provide valid input, then the new Machine will fail to come up and the operator will go degraded, prompting them to resolve the issue, so I think we are ok with this.

That said, AFAIK, we do provision across failure domains, eg on AWS I know when I create a US-East cluster, the installer creates 6 machinesets across all 6 of the failure domains, so maybe it does create the infrastructure

allowing Machine API to create the instances, and then removing their old, manually created control plane instance.
This effectively means migrating their entire control plane onto Machine API before the `ControlPlaneMachineSet` will
take over the management.
We enforce this recommendation so that users can be confident in the Machine `providerSpec` that they have configured
and that Machine API will be able to create valid Machines from the spec before they then start using our automation.

##### Why not to populate the spec for the customer

One idea posed, is to allow the `ControlPlaneMachineSet` to be automatically populated if the spec is left blank when it
is created. This would improve the UX by allowing customers not to have to worry about the spec and allow it to be
inferred from the existing Machines within the cluster. It would also make it easier for managed services to adopt
`ControlPlaneMachineSets` throughout the fleet without having to manually check each cluster.

There are however a few concerns about this idea which means we are planning to not implement this (at least for the
first iteration of the project):

> - It may encourage users to attempt to use `ControlPlaneMachineSets` with UPI clusters which are incorrectly configured

Author: The controller is designed in a way such that, without an existing working set of Control Plane Machines it will degrade and not take any action, we will inform the user via the status of actions they need to take to resolve the issue.

So yes, unless the user has correctly configured a set of Machines, we will not do anything, in the same way the etcd protection does.

of updated configuration to the Control Plane Machines.

### User Stories

- As a cluster administrator of OpenShift, I would like to be able to safely and automatically vertically resize my
  control plane as and when the demand on the control plane changes
- As a cluster administrator of OpenShift, I would like to be able to automatically recover failed Control Plane
  Machines
- As a cluster administrator of OpenShift, I would like to be able to make changes to the configuration of the
  underlying hardware of my control plane and have these changes safely applied using immutable hardware concepts
- As a cluster administrator of OpenShift, I would like to be able to control rollout of changes to my Control Plane
  Machines such that I can test changes with a single replica before applying the change to all replicas
- (Future work) As a cluster administrator of OpenShift with restricted hardware capacity, I would like to be able to   
  scale down my control plane before adding new Control Plane Machines with newer configuration, notably, my
  environment does not have capacity to add additional Machines during updates

### API Extensions

> We will introduce a new `ControlPlaneMachineSet` CRD to the `machine.openshift.io/v1beta1` API group. It will be based

Reviewer 6: Are *ANY* of the key user stories here applicable to generic sets of nodes and normal machine sets?  It's helpful framing here to describe why this isn't a convention around MachineSets, or a more general controller type (`StatefulMachineSet`).  I *think* you address the justifications in multiple sections, but whenever you introduce a new type you need to clearly and succinctly describe why *this type* vs the other options.  Choosing something specific to Control Plane has advantages, choosing something generic but more powerful than machine set has advantages, but as a design point it needs to be unambiguous what justifies this.

Reviewer 6: Ok.  And ultimately the design pattern for control plane machine set allows you to be opinionated, and the concrete problem you describe is indeed a control plane problem.

Will review the comments.

Reviewer 6: (also, i like control plane machine set because we *won't* have that for hypershift, which means not having the api fundamentally describes the type of cluster as self-hosted or not)

Author: > And ultimately the design pattern for control plane machine set allows you to be opinionated,

I tried to articulate this in the alternative at the bottom, a list of restrictions we are adding to make this control plane specific

> We will introduce a new `ControlPlaneMachineSet` CRD to the `machine.openshift.io/v1beta1` API group. It will be based

Reviewer 7: is v1beta1 tied to the idea that this will be tech preview?  will you use v1 instead if you decide to go straight to GA?

Author: Currently, the entire Machine API is v1beta1, has been since day 0 and I think the intention was always to evolve it depending on where we got with the Cluster API decision. I was thinking to add this to v1beta1 for a couple of reasons, firstly because it then matches the rest of the API which it is relying on (ie if we bumped Machine API to v1 later, we might need to bump this as well to accommodate whatever changes that bump incurs), but also, because we are talking about future iterations where we this supports Cluster API as well, so perhaps that would also want to be another bump of the API down the line. I'll add some notes on this to the enhancement

When the `ControlPlaneMachineSet` controller observes that a newly created Machine is ready, it will delete the old
Control Plane Machine signalling to the etcd operator to switch the membership of the etcd cluster between the old and
new instances.
Once the membership has been switched, the Machine deletion hook will be removed on the old Machine, allowing it to be
removed by the Machine controller in the normal way.

#### Removing/disabling the ControlPlaneMachineSet

As some users may want to remove or disable the ControlPlaneMachineSet, a finalizer will be placed on the
`ControlPlaneMachineSet` to allow the controller to ensure a safe removal of the ControlPlaneMachineSet, while leaving
the Control Plane Machines in place.

Notably it will need to ensure that there are no owner references on the Machines pointing to the ControlPlaneMachineSet
instances. This will prevent the garbage collector from removing the Control Plane Machines when the
`ControlPlaneMachineSet` is deleted.

If users later wish to reinstall the `ControlPlaneMachineSet`, they are free to do so.

#### Installing a ControlPlaneMachineSet within an existing cluster

> When adding a ControlPlaneMachineSet to a new cluster, the end user will need to define the `ControlPlaneMachineSet`

Reviewer 7: ```suggestion
When adding a ControlPlaneMachineSet to an existing cluster, the end user will need to define the `ControlPlaneMachineSet`
```

resource by copying the existing Control Plane Machine ProviderSpecs.
Once this is copied, they should remove the failure domain and add the desired failure domains to the `FailureDomains`
field within the ControlPlaneMachineSet spec.

To ensure adding a `ControlPlaneMachineSet` to the cluster is safe, we will need to ensure via a validating webhook

> that the replica count defined in the spec is consistent with the actual size of the control plane within the cluster.

Reviewer 7: is it too difficult to let them create an empty controlplanemachineset and then populate it properly for them?

Author: This was previously discussed (https://github.com/openshift/enhancements/pull/1008#discussion_r790866005) but I'm not sure it made it into the enhancement. My main concern here is that there are many scenarios where the config may have changed or may have been broken in some way.

For example, if a customer resized their control plane out of band to add more capacity, then forgot to update the Machine specs to represent the actual state. Now if we infer from the Machines what their configuration is, we may then resize their control plane back down to the previous size, possibly causing a cluster failure as the smaller instances cannot cope with the load on them. That's then us breaking their cluster through automation. If they were the one to come up with the configuration, that's on them for breaking the configuration.

If we were to go down the route of auto populating this, I think we would need some sort of check box that the customer acknowledges the inferred configuration and says yep, go ahead, make that the actual state in case it's not, I accept the responsibility. Any idea how we might do that? I can add it as a future improvement to the project (I don't want to extend the scope right now, I think it's already stretching what we can achieve in 4.11)

Reviewer 7: if ControlPlaneMachineSet had an active/inactive field (which i think i saw some need for elsewhere in the EP while i was reading it) you could create the ControlPlaneMachineSet in an Inactive state, and the admin can review what it is going to do before marking it active/enforced.

Reviewer 7: (and yes, keeping this out of scope for 4.11 sounds perfectly reasonable)

Reviewer 8: I am very excited to see this enhancement and look forward to testing it out.  I assume this means the limitation of 3 control plane nodes is being removed as well but did not see it explicitly listed.

I second the idea of an active/inactive flag to satisfy the potential issue of out of band changes to the control plane.  Put that reconciliation back in the admin.

Author: > if ControlPlaneMachineSet had an active/inactive field

This was mentioned in a previous conversation though I didn't think it had made it in. We have been using `paused` in the upstream Cluster API community so would likely copy that pattern, but it's effectively the same

> I assume this means the limitation of 3 control plane nodes is being removed as well but did not see it explicitly listed.

This is listed as a non-goal, we will only support vertical scaling at first and not horizontal scaling

> I second the idea of an active/inactive flag to satisfy the potential issue of out of band changes to the control plane. Put that reconciliation back in the admin.

You can also effectively do this by using the OnDelete strategy, which won't take action until a user deletes one of the existing Control Plane Machines

required to run at a higher run-level than Machine API.

In UPI or misconfigured clusters, a user adding a `ControlPlaneMachineSet` will result in a degraded cluster.
Users will need to remove the invalid ControlPlaneMachineSet resource to restore their cluster to a healthy state.

We do not recommend that UPI users attempt to adopt their control plane instances into Machine API due to the high
likelihood that they cannot create an accurate configuration to replicate the original control plane instances. This
limitation also limits the `ControlPlaneMachineSet` operator to an IPI only operator.

#### Naming of Control Plane Machines owned by the ControlPlaneMachineSet

In an IPI OpenShift cluster, the Control Plane Machines are named after the cluster ID followed by an index, for example
`my-openshift-cluster-abcde-master-0`. As we will need to create additional Machines during replacement operations, we
cannot reuse the names of the existing Machines. Instead we will use the `generateName` field within the metadata to
create a random suffix for the new Machines, replacing the index of current Machines. For example, Machines created

> by a `ControlPlaneMachineSet` may have a name such as `my-openshift-cluster-abcde-master-fghij`.

Reviewer 7: can't you generate a new random string to include, but still append the appropriate index suffix to the new machine name?

it may not be strictly critical but it is user-friendly

Author: We definitely could do this, though with other Machines we tend not to, mainly because we try to stick to the cattle not pets idea, however, for control plane, I think you're right, it might be nice to see the 0, 1, 2 indexes still, will update

As users can today replace their Machines with differently named Machines (eg as part of a recovery process), we do not
expect other components to be relying on the index of the Machines for any function within OpenShift, but we may
discover during testing something that is currently unknown to us. If we discover issues with other components relying
on a particular naming scheme for the Control Plane Machines, we will need to solve this issue within the other
component.

#### Delivery via the core OpenShift payload

This new operator will form part of the core OpenShift release payload for clusters installed on/upgrading to OpenShift
4.11. While there is a movement within OpenShift to stop adding new components to the release payload, we believe that

> this new operator should be added to the core for the following reasons:

Reviewer 7: i could see this being a candidate for day0 OLM operators in the future, but yes for now i think you make a good case for it being in the payload.

I do think it would be reasonable to have it be an optional capability, however, per https://issues.redhat.com/browse/OCPPLAN-7589 and https://github.com/openshift/enhancements/blob/master/enhancements/installer/component-selection.md

Author: Yep, this seems reasonable, added a note and a link to the component-selection enhancement saying this can be made optional via that mechanism

managing OpenShift Dedicated (OSD), Red Hat OpenShift on AWS (ROSA), and Azure Red Hat OpenShift (ARO).

The procedure to resize a Control Plane today is lengthy and very involved. It takes a significant amount of time for
an OpenShift expert to perform. We also document this process for our end users, however due to the complexity of the
procedure, there is discomfort in recommending this procedure to end users who may not be as familiar with the product.

To ensure the long term usability of OpenShift, we must provide an automated way for users to scale their Control Plane
as and when their cluster needs additional capacity.

As HyperShift adoptions grows, HyperShift will solve the same issue by running hosted Control Planes within management
clusters. However, HyperShift is not a suitable product for all OpenShift requirements (for example the first management
cluster) and as such, we see that HA clusters will continue to be used and we must solve this problem to allow the
continued adoption of HA clusters.

### Goals

* Provide a "best practices" approach to declarative management of Control Plane Machines

> * Allow users to "1-click" scale their Control Plane to large or smaller instances

Author: For compute nodes we don't currently support this. There's a section at the bottom that talks about the idea of a `StatefulMachineSet` where we discuss this a bit more in detail. The TL;DR is that we don't think a `StatefulMachineSet` is warranted, but, if we want the idea of 1 click changes, we should look into the `MachineDeployment` concept from Cluster API. Originally it was decided we didn't want this within OpenShift but perhaps we want to reconsider that decision.

It was our intention to actually borrow a lot of the `MachineDeployment` implementation/use it as inspiration for the logic within this new controller

> * Allow users to "1-click" scale their Control Plane to large or smaller instances

Author: I guess the naming is what the issue is here, if it were for example, a ControlPlaneMachineDeployment, this would make the distinction more obvious as hopefully that would hint at its parallel being a MachineDeployment which we don't (yet) have.

I think we will just have to make sure in our documentation efforts that we are clear on what is the expectation here

    // Replicas defines how many Control Plane Machines should be
    // created by this ControlPlaneMachineSet.
    // This field is immutable and cannot be changed after cluster
    // installation.
    // The ControlPlaneMachineSet only operates with 3 or 5 node control planes,
    // 3 and 5 are the only valid values for this field.
    // +kubebuilder:validation:Enum:=3;5
    // +kubebuilder:default:=3
    // +kubebuilder:validation:Required
    Replicas *int32 `json:"replicas"`

    // Strategy defines how the ControlPlaneMachineSet will update
    // Machines when it detects a change to the ProviderSpec.
    // +kubebuilder:default:={type: RollingUpdate}
    // +optional
    Strategy ControlPlaneMachineSetStrategy `json:"strategy,omitempty"`

    // FailureDomains is the list of failure domains (sometimes called
    // availability zones) in which the ControlPlaneMachineSet should balance
    // the Control Plane Machines.
    // This will be injected into the ProviderSpec in the
    // appropriate location for the particular provider.
    // This field is optional on platforms that do not require placement
    // information, eg OpenStack.
    // +optional

>     FailureDomains []string `json:"failureDomains,omitempty"`

Reviewer 2: Is this field mutable?

What type of balancing will be done on the failure domains? For example, if I start with 1 failure domain and 3 replicas, then I add 3 more failure domains, can I expect the control planes to be re-distributed across the 3 failure domains? Similarly, if I start with 3 failure domains and later reduce down to one failure domain, can I expect the control plane machines to be re-created so that they are all in the single failure domain?

If the user wants to add failure domains that were not selected at install-time, will we provide instructions on how the user is expected to create the network infrastructure to support the new failure domains? Presumably there will not be an in-cluster operator that will do it for them.

Author: > Is this field mutable?

Initially we weren't planning to make it immutable, but did want to restrict, on creation, that the list matches the existing failure domains within the control plane machines. 

> What type of balancing will be done on the failure domains? For example, if I start with 1 failure domain and 3 replicas, then I add 3 more failure domains, can I expect the control planes to be re-distributed across the 3 failure domains? Similarly, if I start with 3 failure domains and later reduce down to one failure domain, can I expect the control plane machines to be re-created so that they are all in the single failure domain?

We have a use case where in vSphere, currently zones doesn't exist as a concept, but in 4.11 the SPLAT team are looking to introduce this. We don't want to stop customers from migrating their no-zone setup to multizone, and in theory this controller should be able to do that.

> If the user wants to add failure domains that were not selected at install-time, will we provide instructions on how the user is expected to create the network infrastructure to support the new failure domains? Presumably there will not be an in-cluster operator that will do it for them.

In the case above, we are expecting users to be given instructions for setting up the zone infrastructure and then being able to expand into the zones from there. I imagine we will do something similar for other platforms should customers wish to do so, however, for now we aren't planning to allow vertical scaling, so adding extra failure domains won't do anything, changing one of the values would cause a redeploy/rolling update

Due to the nature of the actions performed by this controller, we believe it is important that it have its own
`ClusterOperator` in which to report its current status.

The `ControlPlaneMachineSet` CRD will be limited to a singleton within a standard OpenShift HA cluster, the only
allowed name will be `cluster`. This matches other high level CRD concepts such as the `Infrastructure` object.
Due to this limitation, the CRD will be cluster scoped.

The behaviour of such a controller is complex, and as such, various features of the controller and scenarios are
outlined in the details below.

#### Desired number of Machines

At present, the only source of the installed control plane size within OpenShift clusters exists within the
`cluster-config-v1` ConfigMap in the `kube-system` namespace.
This ConfigMap has been deprecated for some time and as such, should not be used for scale in new projects.

Due to this limitation, we will need to have a desired replicas field within the ControlPlaneMachineSet controller.
As we are not planning to tackle horizontal scaling of the control plane initially, we will implement a validating
webhook to deny changes to this value once set.

> For new clusters, the installer will create (timeline TBD) the `ControlPlaneMachineSet` resource and will set the value

Reviewer 2: The installer will continue to create the `Machine` resources for the control plane in addition to the `ControlPlaneMachineSet`, right?

Author: Yes that was the plan, for now, when the `ControlPlaneMachineSet` is created, it is expected that the Machines will already exist, it will go degraded if there are no Control Plane Machines. The CPMS will then take over management of the Machines that the installer creates

number of existing Control Plane Machines matches the replica count set within the `ControlPlaneMachineSet`.
This will prevent end users from attempting to horizontally scale their control plane during creation of the
`ControlPlaneMachineSet` resource.

In the future, once we have identified risks and issues with horizontal scale, and mitigated those, we will remove the
immutability restriction on the replica field to allow horizontal scaling of the control plane between 3 and 5 replicas
as per current support policy for control plane sizing.

#### Selecting Machines

The ControlPlaneMachineSet operator will use the selector defined within the CRD
to find Machines which it should consider to be within the set of control plane Machines it is responsible for managing.

This set should be the complete control plane and there should be a 1:1 mapping of Machines in this set to the control
plane nodes within the cluster.

If there are any control plane nodes (identified by the node role) which do not have a Machine, the operator will mark
itself degraded as this is an unexpected state.

> No further action will be taken until the unknown node has either been removed from the cluster or a Machine has been

Author: It means that the CPMS will just stop taking actions, it will calculate the status of the objects but will not create/delete any Machines

manually created for it.

#### Providing high availability/fault tolerance within the control plane

Typically within OpenShift, Machines are created within a MachineSet.
MachineSets have a single `ProviderSpec` which defines the configuration for the Machines created by the MachineSet.
The failure domain (sometimes called availability zone) is a part of this provider spec and as such, defines that all
Machines within the Machineset share a failure domain.

This is undesirable for Control Plane Machines as we wish to have them spread across multiple availability zones to

> reduce the likelihood of datacenter level faults degrading the control plane.

Reviewer 9: How is this different from compute nodes, which customers also want to be spread across multiple availability zones, and where we meet this need by creating multiple machine sets with one for each failure domain?

Author: It's not any different, but this is particularly undesirable for the Control Plane IMO. 

With workloads, they are the customers responsibility and it is up to them to spread across multiple failure domains. The Control Plane is more in our hands so we should endeavour to make this more visible/obvious and implement the best practice here.

Reviewer 9: Can these guard rails belong somewhere else? For example, the Installer could determine to provision multiple machinesets for control plane just like we currently do for compute. I strongly advise against duplicating code just to enforce "best practice" as those code paths are eventually going to diverge and provide inconsistent user experience.

Author: I've tried to highlight this within the alternatives section, but to summarise, there have been previous iterations at how we might achieve this using just regular MachineSets. With each previous iteration there were a huge number of risks and hacks that were going to be needed to make this safe, for example, what happens when a user deletes a machineset owning a control plane machine, the machine goes away and they lose a control plane member, is that something we want to allow customers to do? There are other examples but they're already documented. Having spoken with a few people, while I agree, the duplication is not ideal, there are also a lot of good reasons to have a separate, specific control plane management which I think outweigh the negative of duplication here

| OpenStack | .availabilityZone           |

Note that on some platforms (eg OpenStack) the failure domain field is optional, as such, the `FailureDomains` field
must also be optional.

###### Failure Domains on vSphere

As of OpenShift 4.10, there is no concept of a zone within vSphere and as such we would expect that the `FailureDomains`
would be omitted for vSphere.

However, [there is future work](https://github.com/openshift/enhancements/pull/918) to include zone support for vSphere
within 4.11. Once this is implemented, the Machine API provider for vSphere will understand the concept of zones,
defined within the `Infrastructure` resource.
The `ControlPlaneMachineSet` will then be able to balance Machines across the different zones by setting the new zone
information within the provider specs as it does on other platforms.

#### Ensuring Machines match the desired state

To ensure Machines are up to date with the desired configuration, the `ControlPlaneMachineSet` controller will leverage
a similar pattern to the workload controllers.
It will hash the template and compare the hashed result with the spec of the Machines within the managed set.

> As we expect the failure domain to vary across the managed Machines, this will need to be omitted before the hash can

Reviewer 2: There are more fields than just the failure domain that will vary across the managed Machines. For example, on AWS, the `Subnet` field will vary based upon which failure domain is chosen.

Author: We had a few discussions about this and how different platforms handle this kind of difference, I think AWS didn't even cross our minds, but the likes of vSphere and OpenStack etc did.

I will need to revisit this topic and see what we can come up with. There has been talk of `FailureDomain` CRDs (CAPI do this for vSphere) that define all of the context required to create a Machine within a failure domain, the alternative is that we do something inline, for example

```
failureDomains:
- name: my-made-up-fd-1
  value:
    availabilityZone: us-east-1a
    subnet: my-subnet-1a
- name: my-made-up-fd-2
  value:
    availabilityZone: us-east-1b
    subnet: my-subnet-1b
```
With which we then overlay the information onto the shared providerSpec.

The other alternative is that they provide 1 providerSpec per failure domain but that isn't great from a UX perspective.

Author: An alternative here, which would be more concrete, would to be to create a discriminated union and have each of the fields that will be able to be overriden within the providerSpec as a concrete field, I think from an API perspective that is likely to be the safer option, this would probably look something like this instead
```
failureDomains:
- aws:
    availabilityZone: us-east-1a
    subnet: my-subnet-1a
- aws:
    availabilityZone: us-east-1b
    subnet: my-subnet-1b
```

Or for Azure:

```
failureDomains:
- azure:
    zone: us-central-1
- azure:
    zone: us-central-2
```

Author: I've pushed an update which redefines the failure domain as a discriminated union, which I think should solve the problem

be calculated.

Should any Machine not match the desired hash, it will be updated based on the chosen update strategy.

##### The RollingUpdate update strategy

The RollingUpdate strategy will be the default. It mirrors the RollingUpdate strategy familiar to users from
Deployments. When a change is required to be rolled out, the Machine controller will create a new Control Plane
Machine, wait for this to become ready, and then remove the old Machine. It will repeat this process until all Machines
are up to date.
During this process the etcd membership will be protected by the mechanism described in a [separate enhancement](https://github.com/openshift/enhancements/pull/943),
in particular it isn't essential that the Control Plane Machines are updated in a rolling fashion for etcd sakes, though

> to avoid potential issues with the spread of etcd members across availability zones during update, the
> `ControlPlaneMachineSet` will perform a rolling update zone by zone.

Reviewer 2: ```suggestion
to avoid potential issues with the spread of etcd members across failure domains during update, the
`ControlPlaneMachineSet` will perform a rolling update domain by domain.
```

At first, we will not allow any configuration of the RollingUpdate (as you might see in a Deployment) and will pin the
surge to 1 replica. We may wish to change this later once we have more data about the stability of this operator and
the etcd protection mechanism it relies on.

We expect this strategy to be used in most applications of the `ControlPlaneMachineSet`, in particular it can only not

> be used in environments where capacity is very limited and a surge of any control plane capacity is unavailable.

Reviewer 2: What does it mean to say "it can only not be used"? That language implies to me that we are placing some explicit restrictions, as opposed to just listing one scenario where the RollingUpdate strategy will not be practical.

Author: Yeah this is poorly worded, will reword. I think I just wanted to highlight that this won't be applicable to all environments

#### Protecting etcd quorum during update operations

With the introduction of the [etcd protection enhancement](https://github.com/openshift/enhancements/pull/943), the
ControlPlaneMachineSet does not need to observe anything etcd related during scaling operations.
In particular, it is the etcd operators responsibility to add Machine deletion hooks to prevent Control Plane Machines
from being removed until they are no longer needed.
When the `ControlPlaneMachineSet` controller observes that a newly created Machine is ready, it will delete the old
Control Plane Machine signalling to the etcd operator to switch the membership of the etcd cluster between the old and
new instances.
Once the membership has been switched, the Machine deletion hook will be removed on the old Machine, allowing it to be
removed by the Machine controller in the normal way.

#### Removing/disabling the ControlPlaneMachineSet

As some users may want to remove or disable the ControlPlaneMachineSet, a finalizer will be placed on the
`ControlPlaneMachineSet` to allow the controller to ensure a safe removal of the ControlPlaneMachineSet, while leaving
the Control Plane Machines in place.

> Notably it will need to ensure that there are no owner references on the Machines pointing to the ControlPlaneMachineSet

Reviewer 2: It is not clear to me why we would even want an owner reference from the `ControlPlaneMachineSet` to the `Machines`. What is the use case for that?

Author: Having the owner reference signals to other controllers in the cluster that there is an owner for the Machine, in which case this changes the logic for those controllers.

Having an owner implies that if you change the object, it may be put back if your change conflicts with the desired spec (eg if a user goes and modifies the Machine, they will cause a rollout and for that Machine to be replaced, the owner will hint to them they shouldn't be modifying it I hope). 

It also implies that the object would be replaced should it be deleted, it will be recreated. This logic is used within the MachineHealthCheck, so adding an Owner Ref to the CPMS managed Machines then enables MHC to work with Control Plane Machines

instances. This will prevent the garbage collector from removing the Control Plane Machines when the
`ControlPlaneMachineSet` is deleted.

If users later wish to reinstall the `ControlPlaneMachineSet`, they are free to do so.

#### Installing a ControlPlaneMachineSet within an existing cluster

When adding a ControlPlaneMachineSet to a existing cluster, the end user will need to define the
`ControlPlaneMachineSet` resource by copying the existing Control Plane Machine ProviderSpecs.
Once this is copied, they should remove the failure domain and add the desired failure domains to the `FailureDomains`
field within the ControlPlaneMachineSet spec.

To ensure adding a `ControlPlaneMachineSet` to the cluster is safe, we will need to ensure via a validating webhook
that the replica count defined in the spec is consistent with the actual size of the control plane within the cluster.

> We will also validate that the failure domains align with those within the cluster already, this will prevent users

Reviewer 2: How is this validation going to be done? Will you inspect cloud resources? Or will you limit the failure domains to just the those in use by active machines at the time that the CR is created?

Author: The intention is to do this via a ValidatingWebhook, looking at the active machines in the cluster at the time the CR is created, will need to identify whether this adds significant latency to that API create request or not, hopefully it doesn't.

from accidentally migrating their Control Plane Machines from multiple availability zones to a single zone.

If no Control Plane Machines exist, or they are in a non-Running state, the operator will report degraded until this
issue is resolved. This creates a dependency for the `ControlPlaneMachineSet` operator on the Machine API. It will be
required to run at a higher run-level than Machine API.

> In UPI or misconfigured clusters, a user adding a `ControlPlaneMachineSet` will result in a degraded cluster.

Reviewer 2: How is the determination going to be made that the cluster is UPI? UPI is an installation style but results in the same cluster as IPI.

Author: I'm using UPI as shorthand here so will update to spell it out, What I actually want to say is in a cluster where it does not have valid Control Plane Machines and Nodes, the cluster will degrade.

If a user is willing to create Control Plane Machines in their UPI cluster and can get them to a Running phase, they can then add a CPMS to take over management, but we don't intend to do that initial bootstrap for them

Users will need to remove the invalid `ControlPlaneMachineSet` resource to restore their cluster to a healthy state.

> We do not recommend that UPI users attempt to adopt their control plane instances into Machine API due to the high

Reviewer 2: I do not agree with this sentiment. This will just add another tick to making UPI a lower-class cluster. I do not think that the likelihood is high that the user cannot create an accurate configuration. I would expect in most cases that the configuration would be pretty close to what the IPI would have created already.

Author: I'm going to reword this section where it talks about UPI, you're right, we shouldn't be making this distinction, I just want to make sure it's clear that users must ensure they have "Running" Control Plane Machines before they create the CPMS

  controlled way

In most Kubernetes environments and cloud applications, the IP address of the host does not matter and is not required
to be stable. The application layer networking means that the host IP address is insignificant to the functionality of
the cluster. The only use case we can consider where static IPs over a given set of hosts are required would be in a
scenario where you have an external load balancer that requires reconfiguration if the host IPs change. However, for
this example, it would most likely be more cloud-native to implement an operator that could reconfigure the load
balancer on changes rather than trying to keep the IP addresses of the hosts static. Additionally, there are already
projects tackling IPAM within Kubernetes which may resolve this issue without having to make pets of the Machines.

For storage, we expect most users to use persistent volumes which can be, in most environments attached to multiple
hosts, whether that be abstracted away as a cloud provider service (eg AWS EFS) or as an iSCSI storage network Within
a datacenter. In certain applications these network storage provisions may not be suitable however and you may need

> access to a local disk or volume. In cloud environments this doesn't apply, in virtualized environments the the local

Reviewer 10: minor nit
```suggestion
access to a local disk or volume. In cloud environments this doesn't apply, in virtualized environments the local
```

    // +optional
    Conditions []metav1.Condition `json:"conditions,omitempty"`
}

// FailureDomain represents the different configurations required to spread Machines
// across failure domains on different platforms.
// +union
type FailureDomain struct {
  // Platform identifies the platform for which the FailureDomain represents
  // +kubebuilder:validation:Enum:="aws";"azure";"gcp";"openstack"
  // +unionDiscriminator
  // +optional

>   Platform configv1.PlatformType `json:platform,omitempty`

Reviewer 11: I realise this forms part of a union discriminator pattern (although I couldn't find exactly where that's defined), but it seems like a strangely high level value to be buried so deeply in this structure.

Author: You might find [this doc](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/1027-api-unions/README.md) helpful as it talks about how they work

This field is optional, and I don't expect most users to set it. The value of this field, IIUC, must match the json name of the struct that is non nil within the union, but as long as only one struct is non-nil, it's not required.
This field is purely informational as to which struct in the union is intended to be used

  // AWS configures failure domain information for the AWS platform
  // +optional

>   AWS *AWSFailureDomain `json:aws,omitempty`

Reviewer 11: Enumerating all the things... Is there really no other way?

Incidentally, I do really like the idea of a structure containing infrastructure defaults specific to an AZ. I suspect you'll soon want to reference these from MachineSet/Machine.

Author: > Enumerating all the things... Is there really no other way?

The alternative is to have a list of raw extension and allow users to put any yaml in here that we want, and then writing something to overlay that over the providerSpec that they provide us.

I think it's preferable to have this structured and control what we will allow them to vary per failure domain.

The alternative is that information such as this is move centrally out of the Machine provider specs (as is done in CAPI), but then that's a major change to MAPI which we can't afford to make right now.
